test/core/test_exitcode.py

```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2024 Thomas Krug
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import pytest
+
+import test
+
+minimal_config = pytest.mark.parametrize("manager", [test.helpers.BareConfig], indirect=True)
+
+
+@minimal_config
+def test_exitcode_default(manager):
+    manager.c.shutdown()
+
+    manager.proc.join(10)
+    exitcode = manager.proc.exitcode
+
+    assert exitcode == 0
+
+
+@minimal_config
+def test_exitcode_explicit(manager):
+    code = 23
+
+    manager.c.shutdown(code)
+
+    manager.proc.join(10)
+    exitcode = manager.proc.exitcode
```

Comment:
My guess is that the "same thing" is happening: because qtile exits here without doing that much, it writes down "i didn't have much coverage", and then somehow that gets interpreted as coverage for the whole thing.

Unfortunately, I don't know much about how the coverage statistics actually work, so I'm not sure how to advise you to proceed. The best thing I can think is to try and unset some env vars or unload some injected modules from this copy of the manager... maybe via manager.c.eval()? But I don't know what those env vars / injected modules would be, and I suspect they would be subject to change. Maybe coveralls has a special env var you can set for "don't report coverage"?

Reply:
After some testing I believe the issue is os._exit() leads to the installed coverage atexit handler not being run.
But I have to use the _exit variant because otherwise we would get a loop on shutdown and I need to explicitly call an exit method to set the exitcode.

I would have to mock the handler method removing os._exit which would reduce the test to absurdity.
My suggestion is to remove the test case or just test the command setting the exitcode within the lifecycle intsance.

Reply:
Ok first up sorry for the noise, I will squash the commits.
What I came up with is that always calling os._exit lead to not executing the atexit handlers (pytest, cov) and thereby the coverage reports were missing for a lot of testcases.
This is fixed by the if statement in lifecycle.

On the test side I went back to starting own Qtile instances, which seems to work fine.
I just have a problem when creating Qtile with the wayland backend, which is why I currently skip the test for wayland.